### ISSUE TITLE
confused abou training method

### ISSUE THREAD

**Issue by developer239** [wagenaartje/neataptic#86]
I wanted to make the neural network to learn how to count using number: 1, 2, 3, 4, 5, 6, 7, 8, 9.

I normalized my inputs to 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9.

Then I wrote really simple training program that would teach the net how to add 1 + 2 (0.1 + 0.3).

```
const architect = require('neataptic').architect


const myTrainingSet = [
  { input: [0.1, 0.2], output: [0.3] },
  { input: [0.2, 0.1], output: [0.3] }
];

myNetwork = architect.Perceptron(2, 3, 1);

myNetwork.train(myTrainingSet, {
  log: 1,
  error: 0.01,
  iterations: 1000,
  rate: 0.3
});

console.log(myNetwork.activate([0,0]));
console.log(myNetwork.activate([1,1]));
console.log(myNetwork.activate([0.1,0.2]));
```

The problem is that this logs:

```
[ 0.3717501873608793 ]
[ 0.3695919770977549 ]
[ 0.37142744367869446 ]
```

It basically logs 0.3 for every input. Can some please explain what have I done wrong? :)

**Comment by wagenaartje** [comment]
Neural networks need to learn through training data! You have just provided two samples. A neural network needs more samples to learn such operations. 

Coincidentally, your samples always have `0.3` as output -> now the network thinks that it should always output `0.3`. 

In order to get more sophisticated outputs from neural networks, they should be fed with enough training points so that they actually notice relationships between inputs and outputs.

<hr>

I created a working example [here](https://jsfiddle.net/a8qpab71/)! It creates 1000 samples of addition, the result may never be higher than 10 (so `1` for the neural network). So you can't add `0.5` and `0.6` for example.

If you have anymore questions, let me know!

**Comment by D-Nice** [comment]
The training set for one is far too limited. More data is always better, especially diverse data. The above trainingset is almost like feeding it a single one, as the outputs are the same, you want as diverse a range of outputs as possible too.

However, it still won't necessarily give you what you're looking for, outputs will be normalized between 0 and 1 as well, so you won't be able to do anything with a sum greater than 10 (1.0 true output). 

You may wish to try the NEAT algorithm, which still requires a lot of data, but from my experience, has no bounds on the input/output so no need to normalize, and even the output isn't normalized. 

You also mentioned it wanting to learn how to count, this should be easily achievable via an RNN.

**Comment by developer239** [comment]
@wagenaartje @D-Nice I eventually want to create NEAT bot playing computer game (similar to the IAmDinosaur) It is quite hard though because I have very little knowledge and there are not many comprehensive tutorials. This is a great start. Thanks a lot. :)

**Comment by developer239** [comment]
@wagenaartje @D-Nice Hey I just spent few more hours playing with javascript and neural networks but I couln't find what I was looking for.anywhere.

Your library is by far the best. However your tutorials are way too compliacted. I was going through source code of agar-io and the follower thing but there is way too much stuff going on.

I have two questions if that is ok (because people on stackoverflow don't really help anymore...)

1. What is the difference between `train` and `evolve` methods? I mean to me it looks like it is the same thing. I still have to have my own training set in order to get some result.

2. How do I use the NEAT functionality? Could you please show me simple example of adding up to 10 using neat instead of train / evolve methods? All of this stuff is extremely interesting but I am afraid that it will be quite hard to get into it.

This is really stupid example of how I thought that I am supposed to use NEAT to teach XOR.

```
const { Neat, methods } = require('neataptic')


function fitnessFunction(genome) {
  // something like
  let integer1 = Math.floor(Math.random() * 10)
  let integer2 = Math.round(Math.random() * (10 - integer1))
  const output = genome.activate(integer1 / 10, integer2 / 10)
  // Now I would evaluate the output I would give genome points based on how close it is to real answer
}

async function run() {
  const network = new Neat(2, 1, fitnessFunction, {
    mutation: methods.mutation.ALL,
    mutationRate: 0.3,
    clear: true,
    elitism: 0.1,
  })

  // I have no idea how to use this
  // Is this the same thing as https://wagenaartje.github.io/neataptic/docs/important/evolve/ ?
  // because I already passed the setting to Neat constructor
  await network.evolve()
  
  // This one is also confusing
  // In documentation: Evaluates the entire population by passing on the genome to the fitness function and taking the score.
  // How do I use it then please?
  await network.evaluate()
  
  // What do I do now?
}

run()
```

Thanks a lot. I also went through all the issues and found out about this thing: https://github.com/AlexisTM/FlappyNet But I was not surprised that it does not work. :/ It looked so promising.

**Comment by tylerlindell** [comment]
@developer239 here is a tutorial for how to implement NEAT playing IAmDinosaur (in Spanish): https://www.youtube.com/watch?v=P7XHzqZjXQs

additionally, you can learn to implement NEAT in Unity here: https://goo.gl/wykc7J

**Comment by developer239** [comment]
@tylerlindell That video is basically the thing that got me interested in NEAT. He does not talk about the actual implementation though. But thanks for your interest. :)

I am currently looking at this guys repo: https://www.youtube.com/watch?v=4O61sPaGI50

**Comment by wagenaartje** [comment]
Hi, thankyou for your feedback.

1. The difference is that the `train` function trains the network through regular backpropagation. The `evolve` function trains the network through evolution (with NEAT).

If you have a **fixed** data set, like XOR, MNIST etc. then you can just use the regular `evolve` function. This uses NEAT to develop a network to map the data set. [This](https://jsfiddle.net/a8qpab71/2/) is an example of a neural network learning to do additions up to 10 through neuro-evolution (NEAT).

There is an example where XOR is trained with NEAT first thing on the [Readme](https://github.com/wagenaartje/neataptic).

However, Flappy Bird is a **"dynamic"** data set (there is no data, there is just good or wrong, which is measurable through score). Neural networks will evolve through genetic algorithms, using a customized NEAT setup with Neataptic. These kind of implementations can be seen at the Agar.IO example you mentioned. 

Have you checked [NEAT docs](https://wagenaartje.github.io/neataptic/docs/neat/)? I agree it isn't very clear. I will update it soon and make it clearer.

**Comment by developer239** [comment]
@wagenaartje Thank you for your time I really appreciate it :)

I looked at the agar-io example again after somehow creating my own class in synaptic. This is what created using neataptic. Now it seems to be really easy:

```
const { Neat, methods } = require('neataptic')

let neat
let players
let highestScore

const possibleXorInputs = [
    [0, 0], // 0
    [1, 1], // 0
    [0, 1], // 1
    [1, 0], // 1
]

function round(x, n) {
    return Math.round(x * Math.pow(10, n)) / Math.pow(10, n)
}

function getRandomInt(min, max) {
    return Math.floor(Math.random() * (max - min + 1)) + min
}

function initNeat() {
    neat = new Neat(
        2,
        1,
        null,
        {
            mutation: methods.mutation.ALL,
            elitism: 0.1,
            equal: true,
        }
    )
}

function startEvaluation() {
    players = []
    highestScore = 0

    for (let genome in neat.population) {
        // console.log('for genome')
        genome = neat.population[genome]

        let xorIndex = getRandomInt(0, 1)
        let input = genome.activate(possibleXorInputs[xorIndex])
        // let xorIndex = 2
        let output = round(input, 5)
        
        gnome.score = calculateFitness(input, output)
    }
}

function endEvaluation() {
    console.log('Generation:', neat.generation, '- average score:', neat.getAverage())

    neat.sort()
    let newPopulation = []

    // Elitism
    for (let i = 0; i < neat.elitism; i++) {
        newPopulation.push(neat.population[i])
    }

    // Breed the next individuals
    for (let i = 0; i < neat.popsize - neat.elitism; i++) {
        newPopulation.push(neat.getOffspring())
    }

    // Replace the old population with the new population
    neat.population = newPopulation
    neat.mutate()
}

function evolve() {
    neat.generation++
    startEvaluation()
}

initNeat()

for (let i = 0; i < 1000; i += 1) {
    startEvaluation()
    endEvaluation()
    evolve()
}

```

I am struggling with the fitness function though. I can make it get close to 0 for `[0,0]` or `[1,1]` I can make it get close to 1 fot `[0,1]` or `[1,0]` but I have no idea how to combine all these together. 

I used something like this:

```
function startEvaluation() {
    players = []
    highestScore = 0

    for (let genome in neat.population) {
        // console.log('for genome')
        genome = neat.population[genome]

        let xorIndex = getRandomInt(0, 1)
        let input = possibleXorInputs[xorIndex]
        let output = round(genome.activate(input), 5)

        // close to 0
        if (xorIndex === 0 || xorIndex === 1) {
            if(output === 0) {
                genome.score = 1
            } else {
                genome.score = (1 / Math.abs(output))
            }
        }

        // close to 1
        if (xorIndex === 2 || xorIndex === 3) {
            if (output > 0) {
                if ((Math.abs(output) -1) === 0) {
                    genome.score = 1
                } else {
                    genome.score = (1 / (Math.abs(output) -1))
                }
            } else {
                if ((Math.abs(output) +1) === 0) {
                    genome.score = 0
                } else {
                    genome.score = (1 / (Math.abs(output) +1))
                }
            }
        }
    }
}
```

I am getting there though :D I can't use the fixed data set because I want to create NEAT algorithm playing computer game. I got `node-opencv` compute vision already set up. This is basically last piece of the puzzle.

**Comment by wagenaartje** [comment]
Ahah I understand. You want some example that you could extend into a computer game. What you are doing at the moment is checking how the genome performs on óne sample (if I am correct). You should check all the possible samples.

```js
function startEvaluation() {
    var players = []
    var highestScore = 0

    for (let genome of neat.population) {
        genome.score = 0;

        for (let i = 0; i < possibleXorInputs.length; i++) { // go over each sample
          let input = possibleXorInputs[i];
          let output = genome.activate(input); // don't round, if you round then 0.51 is as good as 1
          
          if (i <= 1) { // output should be 0 for first two samples
            genome.score += 1 - output;
          } else { // output should be 1 for first two samples
            genome.score += output;
          }
        }
       
    }
}
```

**Comment by developer239** [comment]
@wagenaartje Wow this is awesome. I change the settings a little. I set `elitism` to `0.8` and that made all the difference in the world. I also turned `euqal` to `false`. I passed custom `network` so that it is more similar to the one in the simpler example. I set `mutationRate` to `0.1`.

Thank you so much this is awesome.

I will post link to working code when I refactor it a little. Right now it is lot of spaghetti...

**Comment by developer239** [comment]
How cool is this? :D 

![image](https://user-images.githubusercontent.com/14946081/31319872-4db71652-ac6b-11e7-9289-f2f593a5940c.png)

**Comment by wagenaartje** [comment]
Awesome! The main key of machine learning in general is finding the right parameters.

But what I find is very weird is that setting elitism to `0.8` made a difference, as it should actually be an integer. However, elitism of `0.8` is the same as elitism of `1`.  Meaning that the best genome is transferred to the next generation without any modifications. 

Could you try to set `mutationRate` to `0.6` and `elitism` to `5` and see how fast it converges? It should converge faster.

**Comment by developer239** [comment]
Thanks for your help again.

This is my refactored code: https://github.com/developer239/neural-network-playground/tree/master/neatXOR

I can close the issue for now. :)

**Comment by developer239** [comment]
@wagenaartje I missed your last comment. You were right about the elitism. I was just lucky enough to get better result when I changed the elitism. I updated my parameters like you said. It seems to be little better.

I will definitely try to evaluate the out more mathematically in the future though. :D Only after creating somehow custom neat class I realized how important the parameter `error` really is.
